src/Integrals.jl

```diff
@@ -147,5 +147,107 @@ function __solvebp_call(prob::IntegralProblem, alg::VEGAS, sensealg, lb, ub, p;
+    end
+    return grid
+end
+
+@inline myselectdim(y::AbstractArray{T,dims}, d, i) where {T,dims} = selectdim(y, d, i)
+@inline myselectdim(y::AbstractArray{T,1}, _, i) where {T} = @inbounds y[i]
+
+@inline dimension(::Val{D}) where D = D
+function __solvebp_call(prob::IntegralProblem, alg::Trapezoidal{S, D}, sensealg, lb, ub, p; kwargs...) where {S,D}
+    # since all AbstractRange types are equidistant by design, we can rely on that
+    @assert prob.batch == 0
+    # using `Val`s for dimensionality is required to make `selectdim` not allocate
+    dim = dimension(D) 
+    p = p
+    if is_sampled_problem(prob)
+        @assert alg.spec isa AbstractArray "For pre-sampled problems where the integrand is an array, the integration grid must also be specified by an array."
+    end
+
+    grid = construct_grid(prob, alg, lb, ub, dim)
+    
+    err = Inf64
+    if is_sampled_problem(prob)
+        data = prob.f
+        # inlining is required in order to not allocate
+        @inline function integrand(i) 
+            # integrate along dimension `dim`, returning a n-1 dimensional array, or scalar if n=1
+            myselectdim(data, dim, i) 
+        end 
+    else
+        if isinplace(prob)
+            y = zeros(eltype(lb), prob.nout)
+            integrand = i -> @inbounds (prob.f(y, grid[i], p); y)
+        else
+            integrand = i -> @inbounds prob.f(grid[i], p)
+        end
+    end
+
+    firstidx, lastidx = firstindex(grid), lastindex(grid)
+
+    out = integrand(firstidx)
+
+    if isbits(out) 
+        # fast path for equidistant grids
+        if grid isa AbstractRange 
+            dx = grid[begin+1] - grid[begin]
+            out /= 2
+            for i in (firstidx+1):(lastidx-1)
+                out += integrand(i)
+            end
+            out += integrand(lastidx)/2
+            out *= dx
+        # irregular grids:
+        else 
+            out *= (grid[firstidx + 1] - grid[firstidx])
+            for i in (firstidx+1):(lastidx-1)
+                @inbounds out += integrand(i) * (grid[i + 1] - grid[i-1])
+            end
+            out += integrand(lastidx) * (grid[lastidx] - grid[lastidx-1])
+            out /= 2
+        end
+    else # same, but inplace, broadcasted
+        out = copy(out) # to prevent aliasing
+        if grid isa AbstractRange 
+            dx = grid[begin+1] - grid[begin]
+            out ./= 2
+            for i in (firstidx+1):(lastidx-1)
+                out .+= integrand(i)
+            end
+            out .+= integrand(lastidx) ./ 2
+            out .*= dx
+        else 
+            out .*= (grid[firstidx + 1] - grid[firstidx])
+            for i in (firstidx+1):(lastidx-1)
+                @inbounds out .+= integrand(i) .* (grid[i + 1] - grid[i-1])
+            end
+            out .+= integrand(lastidx) .* (grid[lastidx] - grid[lastidx-1])
+            out ./= 2
+        end
+    end
+    return SciMLBase.build_solution(prob, alg, out, err, retcode = ReturnCode.Success)
+end
```

Comment:
This implementation of Trapezoid should be in a separate file trapezoid.jl that implements the method, and is then `include`d into the top level file.